### PR TITLE
feat(zonespec): add a spec for asynchronous tests

### PR DESCRIPTION
**Pull request by juliemr**
This spec is constructed with a done callback and a fail callback,
and waits until all asynchronous tasks are completed before
exiting.

### CODE REVIEW

**Comment by juliemr** [comment]
WIP - still needs tests.

**Comment by juliemr** [comment]
Tests are now added! Please take a look. cc @mhevery @vikerman

**Comment by juliemr** [comment]
This will be updated on top of https://github.com/angular/zone.js/pull/275

File: dist/async-test.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
+/******/ (function(modules) { // webpackBootstrap
```
Comment (mhevery) [inline dist/async-test.js]: please revert the dist folder. I will rebuild when we are ready to cut release

File: karma.conf.js [hunk 2]
```diff
@@ -51,7 +51,7 @@ module.exports = function (config) {
 
     logLevel: config.LOG_INFO,
 
-    browsers: ['Firefox'],
+    browsers: ['Chrome'],
```
Comment (mhevery) [inline karma.conf.js]: was this meant to be here? Maybe a separate SHA?

File: gulpfile.js [hunk 3]
```diff
@@ -108,7 +112,8 @@ gulp.task('build', [
   'build/long-stack-trace-zone.js',
   'build/long-stack-trace-zone.min.js',
   'build/wtf.js',
-  'build/wtf.min.js'
+  'build/wtf.min.js',
+  'build/async-test.js'
 ]);
```
Comment (vikerman) [inline gulpfile.js]: I don't know the full details but just looking at the other specs - Do we also need a async-test.min.js?

File: lib/zone-spec/async-test.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+              }
+            }, 0);
+          });
+        }
+      }
+    }
+
+    // ZoneSpec implementation below.
+
+    name: string = 'asyncTestZone';
```
Comment (vikerman) [inline lib/zone-spec/async-test.ts]: Isn't the name already set in the constructor?

File: lib/zone-spec/async-test.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
+(function() {
+  class AsyncTestZoneSpec implements ZoneSpec {
+    _finishCallback: Function;
+    _failCallback: Function;
+    _pendingMicroTasks: boolean = false;
+    _pendingMacroTasks: boolean = false;
+    _pendingEventTasks: boolean = false;
+    _alreadyErrored = false;
```
Comment (vikerman) [inline lib/zone-spec/async-test.ts]: _alreadyErrored: boolean = false ?

File: test/async-test.spec.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    var atz = Zone.current.fork(testZoneSpec);
+
+    atz.run(function() {
+      setTimeout(() => {
+        finished = true;
+      });
+    });
+  });
+
+  it('should call finish after a setTimeout is done', (done) => {
```
Comment (vikerman) [inline test/async-test.spec.ts]: Repeated?

File: test/async-test.spec.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    atz.run(function() {
+      // Promise rejection does not cause onHandleError to trigger
+      // until the next Macro task, so that there's a chance
+      // for handlers to be added.
+      Promise.reject('my reason');
+    });
+
+  });
+});
```
Comment (vikerman) [inline test/async-test.spec.ts]: The zone spec has explicit logic to test whether micro, macro and event tasks are empty. So should we have tests that mix the task types and wait for all of them to drain?

File: dist/async-test.js (hunk 1 above)
Comment (juliemr) [inline dist/async-test.js]: done.

File: gulpfile.js (hunk 3 above)
Comment (juliemr) [inline gulpfile.js]: Nope, I didn't add the build steps to make a min version, since this will only be for testing.

File: karma.conf.js (hunk 2 above)
Comment (juliemr) [inline karma.conf.js]: Oops, reverted.

File: lib/zone-spec/async-test.ts (hunk 5 above)
Comment (juliemr) [inline lib/zone-spec/async-test.ts]: done.

File: lib/zone-spec/async-test.ts (hunk 4 above)
Comment (juliemr) [inline lib/zone-spec/async-test.ts]: Yes, moved this to just the declaration.

File: test/async-test.spec.ts (hunk 6 above)
Comment (juliemr) [inline test/async-test.spec.ts]: Yup, removed.

File: test/async-test.spec.ts (hunk 7 above)
Comment (juliemr) [inline test/async-test.spec.ts]: Done! Added some. Also, that uncovered some bugs with the event task searching, so I fixed it up. Basically, we never want to wait on event tasks.

**Pull request closed without merging**